### PR TITLE
Update: `no-restricted-{imports,modules}`: add “patterns” (fixes #6963)

### LINKED ISSUE #6963: Allow no-restricted-imports to restrict sub-paths of imports
> The version of ESLint you are using.

3.3.1

> The problem you want to solve.

I have a package that contains some files that are considered to be part of its public API (e.g. "my-package/public/foo.js") which we consume by importing these files from the package directly (e.g. `import foo from 'my-package/public/foo.js';`). This package also contains files that are considered to be private and should not be imported (e.g. "my-package/private/bar.js"). I want to allow people to import the package (`import myPackage from 'my-package';`) or from the public directory, but not from the private directory.

> Your take on the correct solution to problem.

I originally considered [enhancing eslint-plugin-import's no-restricted-paths rule to allow for packages](https://github.com/benmosher/eslint-plugin-import/pull/500) but there @ljharb and @benmosher suggested enhancing the core no-restricted-imports rule instead.

### PR DESCRIPTION
**Pull request by ljharb**
**What is the purpose of this pull request? (put an "X" next to item)**

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[x] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

**What rule do you want to change?**
`no-restricted-imports` and `no-restricted-modules`

**Does this change cause the rule to produce more or fewer warnings?**
More, only when the "patterns" option is set.

**How will the change be implemented? (New option, new default behavior, etc.)?**
New schema - the existing schema of an array of strings will continue to work; an alternative object schema with "paths" will also be accepted. A "patterns" property can be added to this object to opt in to the new behavior.

**Please provide some example code that this change will affect:**

``` js
import pick from 'lodash/pick';
/* or */
var pick = require('lodash/pick');
```

**What does the rule currently do for this code?**
Does not warn unless the exact full path is provided.

**What will the rule do after it's changed?**
Allow a pattern to be provided such that a partial path may be provided.

Fixes #6963.

### CODE REVIEW

File: docs/rules/no-restricted-imports.md [hunk 1]
```diff
@@ -39,9 +54,15 @@ import fs from 'fs';
...
+/*eslint no-restricted-imports: ["error", { "paths": ["cluster"] }]*/
 
-import cluster from ' cluster ';
+import cluster from 'cluster ';
+```
+
+```js
+/*eslint no-restricted-imports: ["error", { "patterns": ["lodash/*"] }]*/
+
+import pick from 'lodash/pick ';
```
Comment (platinumazure) [inline docs/rules/no-restricted-imports.md]: Trailing space in the module path, presumably a typo?

File: lib/rules/no-restricted-imports.js [hunk 2]
```diff
@@ -17,33 +27,55 @@ module.exports = {
...
                 if (node && node.source && node.source.value) {
 
-                    const value = node.source.value.trim();
+                    const importName = node.source.value.trim();
 
-                    if (restrictedImports.indexOf(value) !== -1) {
+                    if (restrictedPaths.has(importName)) {
                         context.report(node, "'{{importName}}' import is restricted from being used.", {
-                            importName: value
+                            importName
```
Comment (platinumazure) [inline lib/rules/no-restricted-imports.js]: While you're in here, can we switch this rule to use new-style reporting? That takes a single object argument and looks like this:

```js
context.report({
    node,
    message: MESSAGE,
    data: { importName }
});
```

File: lib/rules/no-restricted-imports.js [hunk 3]
```diff
@@ -17,33 +27,55 @@ module.exports = {
...
                         });
                     }
+                    if (restrictedPatterns.length > 0 && ig.ignores(importName)) {
+                        context.report(
+                            node,
+                            "'{{importName}}' import is restricted from being used by a pattern.",
+                            { importName }
+                        );
```
Comment (platinumazure) [inline lib/rules/no-restricted-imports.js]: Could this use new-style reporting as well?

File: lib/rules/no-restricted-modules.js [hunk 4]
```diff
@@ -53,36 +76,28 @@ module.exports = {
...
-                        });
+                    // node has arguments and first argument is string
+                    if (node.arguments.length && isString(node.arguments[0])) {
+                        const moduleName = node.arguments[0].value.trim();
+
+                        // check if argument value is in restricted modules array
+                        if (restrictedPaths.has(moduleName)) {
+                            context.report(node, "'{{moduleName}}' module is restricted from being used.", {
+                                moduleName
+                            });
```
Comment (platinumazure) [inline lib/rules/no-restricted-modules.js]: Using new-style reporting here would be awesome as well!

File: lib/rules/no-restricted-modules.js [hunk 5]
```diff
@@ -53,36 +76,28 @@ module.exports = {
...
+                        }
+
+                        if (restrictedPatterns.length > 0 && ig.ignores(moduleName)) {
+                            context.report(
+                                node,
+                                "'{{moduleName}}' module is restricted from being used by a pattern.",
+                                { moduleName }
+                            );
```
Comment (platinumazure) [inline lib/rules/no-restricted-modules.js]: Another opportunity for new-style reporting.

**Comment by ljharb** [comment]
@platinumazure no, `import from 'foo/bar'` is not the same as `import { bar } from 'foo'` in any way, unless the package author has gone out of their way to make it so. If you want to forbid specific named imports from a specific path or pattern (which I hadn't thought of), then I think perhaps a better schema is one that is always a single array of items - and an item can be a string path, `{ path: "some string" }`, `{ pattern: "some pattern/*" }`, and either of the object forms could then (now or in a followup PR) add something to handle named imports. The current schema doesn't really fit well with adding per-path named import blocking. Thoughts?

**Comment by platinumazure** [comment]
If the rule hadn't been handling named imports before now, I don't think it
needs to with this PR. I just wanted clarification, which you have
provided. Thanks!

**Comment by platinumazure** [comment]
I'll champion this. We need one more :+1: from the team (someone who isn't vitorbal or not-an-aardvark) before this can be accepted and (hopefully) merged.

**Comment by ljharb** [comment]
@platinumazure updated! I added mentions and examples to both readmes, as well as tests to both rules.

**Comment by nzakas** [comment]
Just a heads up: we have moved to a new CLA checker on pull requests. Even if you've previously signed our CLA, we will need to you sign the new one. To do so, look at the status checks for **licence/cla** and click the "Details" link. Sorry for the inconvenience.

**Comment by ljharb** [comment]
@vitorbal @ilyavolodin @platinumazure @not-an-aardvark what more is needed on this PR before it can be accepted?

**Comment by ljharb** [comment]
whoops, sorry, disregard my comment - github hadn't refreshed the page

**Comment by ljharb** [comment]
@platinumazure ping! :-)

**Comment by ljharb** [comment]
Thanks everyone! Please let me know what else I should do to get this merged :-)

**Pull request merged**